include/gsl/span

```diff
@@ -351,95 +392,97 @@ namespace details
 } // namespace details
 
 // [span], class template span
-template <class ElementType, std::ptrdiff_t Extent>
+template <class ElementType, std::size_t Extent>
 class span
 {
 public:
     // constants and types
     using element_type = ElementType;
     using value_type = std::remove_cv_t<ElementType>;
-    using index_type = std::ptrdiff_t;
+    using size_type = std::size_t;
     using pointer = element_type*;
+    using const_pointer = const element_type*;
     using reference = element_type&;
+    using const_reference = const element_type&;
+    using difference_type = std::ptrdiff_t;
 
-    using iterator = details::span_iterator<span<ElementType, Extent>, false>;
-    using const_iterator = details::span_iterator<span<ElementType, Extent>, true>;
+    using iterator = details::span_iterator<ElementType>;
+    using const_iterator = details::span_iterator<const ElementType>;
     using reverse_iterator = std::reverse_iterator<iterator>;
     using const_reverse_iterator = std::reverse_iterator<const_iterator>;
 
-    using size_type = index_type;
-
 #if defined(GSL_USE_STATIC_CONSTEXPR_WORKAROUND)
-    static constexpr const index_type extent{Extent};
+    static constexpr const size_type extent{Extent};
 #else
-    static constexpr index_type extent{Extent};
+    static constexpr size_type extent{Extent};
 #endif
 
     // [span.cons], span constructors, copy, assignment, and destructor
     template <bool Dependent = false,
-              // "Dependent" is needed to make "std::enable_if_t<Dependent || Extent <= 0>" SFINAE,
-              // since "std::enable_if_t<Extent <= 0>" is ill-formed when Extent is greater than 0.
-              class = std::enable_if_t<(Dependent || Extent <= 0)>>
+              // "Dependent" is needed to make "std::enable_if_t<Dependent || Extent == 0 || Extent
+              // == dynamic_extent>" SFINAE, since "std::enable_if_t<Extent == 0 || Extent ==
+              // dynamic_extent>" is ill-formed when Extent is greater than 0.
+              class = std::enable_if_t<(Dependent ||
+                                        details::is_allowed_extent_conversion<0, Extent>::value)>>
     constexpr span() noexcept : storage_(nullptr, details::extent_type<0>())
     {}
 
-    constexpr span(pointer ptr, index_type count) noexcept: storage_(ptr, count) {}
+    constexpr span(pointer ptr, size_type count) noexcept : storage_(ptr, count)
+    {
+        if (Extent != dynamic_extent) Expects(count == Extent);
+    }
 
     constexpr span(pointer firstElem, pointer lastElem) noexcept
-        : storage_(firstElem, std::distance(firstElem, lastElem))
-    {}
+        : storage_(firstElem, static_cast<std::size_t>(std::distance(firstElem, lastElem)))
```

Comment:
I think this call to distance when you know the inputs are pointers is an unnecessary debug pessimization. Use `lastElem - firstElem` instead?

Reply:
(Probably occurs elsewhere)

Reply:
will do.